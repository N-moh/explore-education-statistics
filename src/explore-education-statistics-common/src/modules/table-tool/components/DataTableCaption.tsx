import commaList from '@common/lib/utils/string/commaList';
import {
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/full-table/types/filters';
import React, { useState } from 'react';
import { FullTableMeta } from '@common/modules/full-table/types/fullTable';
import ButtonText from '@common/components/ButtonText';
import classNames from 'classnames';

export function generateTableTitle({
  timePeriodRange,
  locations,
  subjectName,
  publicationName,
  expanded,
}: {
  timePeriodRange: FullTableMeta['timePeriodRange'];
  locations: FullTableMeta['locations'];
  subjectName: FullTableMeta['subjectName'];
  publicationName: FullTableMeta['publicationName'];
  expanded?: boolean;
}) {
  let timePeriodString = '';

  if (timePeriodRange.length > 0) {
    const startLabel = timePeriodRange[0].label;
    const endLabel = timePeriodRange[timePeriodRange.length - 1].label;

    timePeriodString =
      startLabel === endLabel
        ? ` for ${startLabel}`
        : ` between ${startLabel} and ${endLabel}`;
  }

  const locationsString = ` in ${commaList(
    locations.length > 10 && !expanded
      ? locations.slice(0, 10).map(location => location.label)
      : locations.map(location => location.label),
  )}`;

  return `Table showing '${subjectName}' from '${publicationName}'${
    !expanded && locations.length > 10
      ? `${locationsString} and ${locations.length} other locations...`
      : locationsString
  }${timePeriodString}`;
}

interface Props {
  id: string;
  timePeriodRange: TimePeriodFilter[];
  locations: LocationFilter[];
  subjectName: string;
  publicationName: string;
}

const DataTableCaption = ({
  id = 'dataTableCaption',
  timePeriodRange,
  locations,
  subjectName,
  publicationName,
}: Props) => {
  const [expanded, setExpanded] = useState<boolean>(false);
  const caption = generateTableTitle({
    timePeriodRange,
    locations,
    subjectName,
    publicationName,
    expanded,
  });

  return (
    <>
      <strong id={id}>{caption}</strong>
      {locations.length > 10 && (
        <ButtonText
          className={classNames('govuk-!-display-block govuk-!-margin-top-2')}
          onClick={() => {
            return expanded ? setExpanded(false) : setExpanded(true);
          }}
        >
          {`${expanded ? 'Hide' : 'View'} full table title`}
        </ButtonText>
      )}
      <ul>
        <li>
          <strong>n/a</strong> - value matching this criteria could not be
          found.
        </li>
        <li>
          <strong>x</strong> - value matching this criteria is suppressed.
        </li>
      </ul>
    </>
  );
};

export default DataTableCaption;
