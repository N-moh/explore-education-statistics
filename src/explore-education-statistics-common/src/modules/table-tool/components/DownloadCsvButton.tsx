import ButtonText from '@common/components/ButtonText';
import cartesian from '@common/lib/utils/cartesian';
import {
  CategoryFilter,
  LocationFilter,
  TimePeriodFilter,
  Filter,
} from '@common/modules/full-table/types/filters';
import { saveAs } from 'file-saver';
import Papa from 'papaparse';
import React from 'react';
import { transformTableMetaFiltersToCategoryFilters } from '@common/modules/full-table/utils/tableHeaders';
import { FullTable } from '@common/modules/full-table/types/fullTable';

interface Props {
  publicationSlug: string;
  fullTable: FullTable;
}

const DownloadCsvButton = ({ publicationSlug, fullTable }: Props) => {
  const { subjectMeta, results } = fullTable;
  const {
    indicators,
    filters: metaFilters,
    timePeriodRange: timePeriods,
    locations,
  } = subjectMeta;

  const filters = transformTableMetaFiltersToCategoryFilters(metaFilters);

  const getCsvData = (): string[][] => {
    const filterColumns = Object.entries(metaFilters).map(
      ([key]) => metaFilters[key].legend,
    );

    const indicatorColumns = indicators.map(indicator => {
      const unit = indicator.unit ? ` (${indicator.unit})` : '';
      return `${indicator.label}${unit}`;
    });

    const columns = [
      'Location',
      'Time period',
      ...filterColumns,
      ...indicatorColumns,
    ];

    const rows = cartesian<Filter>(
      locations,
      timePeriods,
      ...Object.values(filters),
    ).map(row => {
      // TODO: Remove ignore when Prettier stops adding trailing comma to tuple type
      // prettier-ignore
      const [location, timePeriod, ...filterOptions] = row as [
        LocationFilter,
        TimePeriodFilter,
        ...CategoryFilter[]
      ];

      const indicatorCells = indicators.map(indicator => {
        const matchingResult = results.find(result => {
          return Boolean(
            filterOptions.every(filter =>
              result.filters.includes(filter.value),
            ) &&
              result.timePeriod === timePeriod.value &&
              result.location[location.level] &&
              result.location[location.level].code === location.value,
          );
        });
        if (!matchingResult) {
          return 'n/a';
        }
        return matchingResult.measures[indicator.value];
      });

      return [...row.map(column => column.label), ...indicatorCells];
    });

    return [columns, ...rows];
  };

  return (
    <ButtonText
      onClick={() => {
        const data = new Blob([Papa.unparse(getCsvData())], {
          type: 'text/csv;charset=utf-8',
        });

        const filename = `data-${publicationSlug}.csv`;

        saveAs(data, filename, {
          autoBom: true,
        });
      }}
    >
      Download the underlying data of this table (.CSV)
    </ButtonText>
  );
};

export default DownloadCsvButton;
