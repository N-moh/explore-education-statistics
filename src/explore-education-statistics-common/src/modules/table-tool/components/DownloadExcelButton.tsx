import ButtonText from '@common/components/ButtonText';
import { FullTableMeta } from '@common/modules/table-tool/types/fullTable';
import { Dictionary } from '@common/types';
import sum from 'lodash/sum';
import React, { RefObject } from 'react';
import { CellObject, utils, WorkSheet, writeFile } from 'xlsx';

interface Props {
  publicationSlug: string;
  footnotes: FullTableMeta['footnotes'];
  tableRef: RefObject<HTMLElement>;
}

/**
 * Explicitly set column widths for a {@param sheet} using
 * the width of the cell with the most content in it.
 * SheetJs doesn't automatically generate our column widths
 * so the spreadsheet would look quite squished without this.
 */
export function appendColumnWidths(sheet: WorkSheet): WorkSheet {
  const columnWidths = Object.entries(sheet)
    .filter(([key]) => !key.startsWith('!'))
    .reduce<Dictionary<number>>((acc, [key, value]: [string, CellObject]) => {
      const [column] = key.split(/[0-9]+/);

      if (typeof acc[column] === 'undefined') {
        acc[column] = 0;
      }

      let length = 0;

      if (typeof value.v === 'string') {
        // eslint-disable-next-line prefer-destructuring
        length = value.v.length;
      } else if (typeof value.v === 'number') {
        // eslint-disable-next-line prefer-destructuring
        length = value.v.toString().length;
      }

      if (length > acc[column]) {
        acc[column] = length;
      }

      return acc;
    }, {});

  // eslint-disable-next-line no-param-reassign
  sheet['!cols'] = Object.entries(columnWidths)
    .map(([key, value]) => {
      // Convert column letters to a single number so they
      // can be sorted in the way that Excel expects
      const keyHash = sum(key.split('').map(char => char.charCodeAt(0)));

      return [keyHash, value];
    })
    .sort(([a], [b]) => a - b)
    .map(([, value]) => {
      return value ? { wch: value } : {};
    });

  return sheet;
}

/**
 * Append {@param footnotes} to the bottom of a {@param sheet}.
 */
export function appendFootnotes(
  sheet: WorkSheet,
  footnotes: FullTableMeta['footnotes'],
): WorkSheet {
  const [, lastCell] = sheet['!ref']?.split(':') ?? [];

  if (!lastCell) {
    return sheet;
  }

  const [lastColumn, lastRow] = lastCell.split(/([0-9]+)/);
  const footnoteStartRow = parseInt(lastRow, 10) + 2;

  footnotes.forEach((footnote, index) => {
    // eslint-disable-next-line no-param-reassign
    sheet[`A${footnoteStartRow + index}`] = {
      t: 's',
      v: `(${index + 1}) ${footnote.label}`,
    } as CellObject;
  });

  // eslint-disable-next-line no-param-reassign
  sheet['!ref'] = `A1:${lastColumn}${footnoteStartRow + footnotes.length - 1}`;

  return sheet;
}

const DownloadExcelButton = ({
  publicationSlug,
  footnotes,
  tableRef,
}: Props) => {
  // Try to find table element within the ref
  let tableEl: HTMLTableElement | null = null;

  if (tableRef.current) {
    if (tableRef.current.tagName.toLowerCase() === 'table') {
      tableEl = tableRef.current as HTMLTableElement;
    } else {
      tableEl = tableRef.current.querySelector('table');
    }

    // Don't render anything if no table could be found
    if (!tableEl) {
      return null;
    }
  }

  return (
    <ButtonText
      onClick={() => {
        const workBook = utils.table_to_book(tableEl, {
          raw: true,
        });
        const sheet = workBook.Sheets[workBook.SheetNames[0]];

        appendColumnWidths(sheet);
        appendFootnotes(sheet, footnotes);

        writeFile(workBook, `data-${publicationSlug}.xlsx`, {
          type: 'binary',
        });
      }}
    >
      Download table as Excel spreadsheet (XLSX)
    </ButtonText>
  );
};

export default DownloadExcelButton;
