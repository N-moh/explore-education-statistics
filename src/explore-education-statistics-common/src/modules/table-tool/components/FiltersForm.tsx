import CollapsibleList from '@common/components/CollapsibleList';
import {
  Form,
  FormFieldCheckboxSearchSubGroups,
  FormFieldset,
  FormGroup,
} from '@common/components/form';
import useFormSubmit from '@common/hooks/useFormSubmit';
import SummaryList from '@common/components/SummaryList';
import SummaryListItem from '@common/components/SummaryListItem';
import FormCheckboxSelectedCount from '@common/modules/table-tool/components/FormCheckboxSelectedCount';
import { SubjectMeta } from '@common/services/tableBuilderService';
import { Dictionary } from '@common/types';
import createErrorHelper from '@common/validation/createErrorHelper';
import { mapFieldErrors } from '@common/validation/serverValidations';
import Yup from '@common/validation/yup';
import { Formik } from 'formik';
import mapValues from 'lodash/mapValues';
import React, { useMemo } from 'react';
import FormFieldCheckboxGroupsMenu from './FormFieldCheckboxGroupsMenu';
import ResetFormOnPreviousStep from './ResetFormOnPreviousStep';
import { InjectedWizardProps } from './Wizard';
import WizardStepFormActions from './WizardStepFormActions';
import WizardStepHeading from './WizardStepHeading';
import WizardStepEditButton from './WizardStepEditButton';

export interface FormValues {
  indicators: string[];
  filters: Dictionary<string[]>;
}

export type FilterFormSubmitHandler = (values: FormValues) => void;

interface Props {
  initialValues?: {
    indicators: string[];
    filters: string[];
  };
  subjectMeta: SubjectMeta;
  onSubmit: FilterFormSubmitHandler;
}

const formId = 'filtersForm';

const FiltersForm = (props: Props & InjectedWizardProps) => {
  const {
    onSubmit,
    subjectMeta,
    goToNextStep,
    currentStep,
    stepNumber,
    initialValues,
    isActive,
  } = props;

  const errorMappings = [
    mapFieldErrors<FormValues>({
      target: 'indicators',
      messages: {
        QUERY_EXCEEDS_MAX_ALLOWABLE_TABLE_SIZE:
          'A table cannot be returned as the filters chosen can exceed the maximum allowable table size',
      },
    }),
  ];

  const initialFormValues = useMemo(() => {
    // Automatically select indicator when one indicator group with one option
    const indicatorValues = Object.values(subjectMeta.indicators);
    const indicators =
      indicatorValues.length === 1 && indicatorValues[0].options.length === 1
        ? [indicatorValues[0].options[0].value]
        : initialValues?.indicators ?? [];

    const filters = mapValues(subjectMeta.filters, filter => {
      const filterGroups = Object.values(filter.options);

      // Automatically select when only one group in filter, with only one option in it.
      if (filterGroups.length === 1 && filterGroups[0].options.length === 1) {
        return [filterGroups[0].options[0].value];
      }

      if (initialValues?.filters) {
        const filterValues = filterGroups
          .flatMap(group => group.options)
          .map(option => option.value);

        return filterValues.filter(filterValue =>
          initialValues.filters.includes(filterValue),
        );
      }

      return [];
    });

    return {
      filters,
      indicators,
    };
  }, [initialValues, subjectMeta]);

  const stepEnabled = currentStep > stepNumber;
  const stepHeading = (
    <WizardStepHeading {...props} stepEnabled={stepEnabled}>
      Choose your filters
    </WizardStepHeading>
  );

  const handleSubmit = useFormSubmit(async (values: FormValues) => {
    await onSubmit(values);
    goToNextStep();
  }, errorMappings);

  return (
    <Formik<FormValues>
      enableReinitialize
      initialValues={initialFormValues}
      validateOnBlur={false}
      validationSchema={Yup.object<FormValues>({
        indicators: Yup.array()
          .of(Yup.string())
          .required('Select at least one option from indicators'),
        filters: Yup.object(
          mapValues(subjectMeta.filters, filter =>
            Yup.array()
              .of(Yup.string())
              .min(
                1,
                `Select at least one option from ${filter.legend.toLowerCase()}`,
              ),
          ),
        ),
      })}
      onSubmit={handleSubmit}
    >
      {form => {
        const { getError } = createErrorHelper(form);

        if (isActive) {
          return (
            <Form {...form} id={formId} showSubmitError>
              {stepHeading}

              <FormGroup>
                <div className="govuk-grid-row">
                  <div className="govuk-grid-column-one-half-from-desktop">
                    <FormFieldCheckboxSearchSubGroups
                      name="indicators"
                      legend={
                        <>
                          Indicators
                          <FormCheckboxSelectedCount name="indicators" />
                        </>
                      }
                      legendSize="m"
                      hint="Select at least one indicator below"
                      disabled={form.isSubmitting}
                      order={[]}
                      options={Object.values(subjectMeta.indicators).map(
                        group => ({
                          legend: group.label,
                          options: group.options,
                        }),
                      )}
                    />

                    {Object.entries(subjectMeta.filters).length > 0 && (
                      <FormFieldset
                        id="filters"
                        legend="Categories"
                        legendSize="m"
                        hint="Select at least one option from all categories"
                        error={getError('filters')}
                      >
                        {Object.entries(subjectMeta.filters).map(
                          ([filterKey, filterGroup]) => {
                            const filterName = `filters.${filterKey}`;
                            const filterGroupOptions = Object.values(
                              filterGroup.options,
                            );

                            return (
                              <FormFieldCheckboxGroupsMenu
                                key={filterKey}
                                name={filterName}
                                legend={filterGroup.legend}
                                hint={filterGroup.hint}
                                disabled={form.isSubmitting}
                                order={[]}
                                options={filterGroupOptions.map(group => ({
                                  legend: group.label,
                                  options: group.options,
                                }))}
                                open={filterGroupOptions.length === 1}
                              />
                            );
                          },
                        )}
                      </FormFieldset>
                    )}
                  </div>
                </div>
              </FormGroup>

              <WizardStepFormActions
                {...props}
                submitText="Create table"
                submittingText="Creating table"
                onSubmitClick={() => {
                  // Automatically select totalValue for filters that haven't had a selection made
                  Object.keys(form.values.filters).forEach(filterName => {
                    if (
                      form.values.filters[filterName].length === 0 &&
                      subjectMeta.filters[filterName].totalValue
                    ) {
                      form.setFieldValue(`filters.${filterName}`, [
                        subjectMeta.filters[filterName].totalValue,
                      ]);
                    }
                  });
                }}
              />
            </Form>
          );
        }

        return (
          <div className="govuk-grid-row">
            <div className="govuk-grid-column-two-thirds">
              {stepHeading}
              <SummaryList noBorder>
                <SummaryListItem term="Indicators">
                  <CollapsibleList>
                    {Object.values(subjectMeta.indicators)
                      .flatMap(group => group.options)
                      .filter(indicator =>
                        form.values.indicators.includes(indicator.value),
                      )
                      .map(indicator => (
                        <li key={indicator.value}>{indicator.label}</li>
                      ))}
                  </CollapsibleList>
                </SummaryListItem>

                {Object.entries(subjectMeta.filters)
                  .filter(([groupKey]) => !!form.values.filters[groupKey])
                  .map(([filterGroupKey, filterGroup]) => (
                    <SummaryListItem
                      key={filterGroupKey}
                      term={filterGroup.legend}
                    >
                      <CollapsibleList>
                        {Object.values(filterGroup.options)
                          .flatMap(group => group.options)
                          .filter(option =>
                            form.values.filters[filterGroupKey].includes(
                              option.value,
                            ),
                          )
                          .map(option => (
                            <li key={option.value}>{option.label}</li>
                          ))}
                      </CollapsibleList>
                    </SummaryListItem>
                  ))}
              </SummaryList>
            </div>
            <div className="govuk-grid-column-one-third dfe-align--right">
              {stepEnabled && (
                <WizardStepEditButton {...props} editTitle="Edit filters" />
              )}
              <ResetFormOnPreviousStep
                currentStep={currentStep}
                stepNumber={stepNumber}
              />
            </div>
          </div>
        );
      }}
    </Formik>
  );
};

export default FiltersForm;
