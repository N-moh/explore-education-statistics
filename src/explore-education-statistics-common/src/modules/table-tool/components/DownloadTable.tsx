import { Form, FormFieldRadioGroup } from '@common/components/form';
import Button from '@common/components/Button';
import {
  FullTable,
  FullTableMeta,
} from '@common/modules/table-tool/types/fullTable';
import {
  appendColumnWidths,
  appendTitle,
  appendFootnotes,
  getCsvData,
} from '@common/modules/table-tool/components/utils/downloadTableUtils';
import { generateTableTitle } from '@common/modules/table-tool/components/DataTableCaption';
import Yup from '@common/validation/yup';
import { Formik } from 'formik';
import React, { createElement, RefObject } from 'react';
import { utils, writeFile } from 'xlsx';

const handleCsvDownload = (fileName: string, fullTable: FullTable) => {
  const workBook = utils.book_new();
  workBook.Sheets.Sheet1 = utils.aoa_to_sheet(getCsvData(fullTable));
  workBook.SheetNames[0] = 'Sheet1';

  writeFile(workBook, `${fileName}.csv`, {
    type: 'binary',
  });
};

const handleOdsDownload = (
  fileName: string,
  subjectMeta: FullTableMeta,
  tableRef: RefObject<HTMLElement>,
) => {
  const { footnotes } = subjectMeta;

  let tableEl: HTMLTableElement | null = null;

  if (tableRef.current) {
    if (tableRef.current.tagName.toLowerCase() === 'table') {
      tableEl = tableRef.current as HTMLTableElement;
    } else {
      tableEl = tableRef.current.querySelector('table');
    }
  }

  if (!tableEl) {
    return;
  }

  const workBook = utils.table_to_book(tableEl, {
    raw: true,
  });
  const sheet = workBook.Sheets[workBook.SheetNames[0]];

  appendColumnWidths(sheet);
  appendTitle(sheet, generateTableTitle(subjectMeta));
  appendFootnotes(sheet, footnotes);

  writeFile(workBook, `${fileName}.ods`, {
    type: 'binary',
    bookType: 'ods',
  });
};

export type FileFormat = 'ods' | 'csv' | undefined;

interface FormValues {
  fileFormat: FileFormat;
}

interface Props {
  fileName: string;
  fullTable: FullTable;
  headingTag?: 'h2' | 'h3' | 'h4';
  headingSize?: 's' | 'm' | 'l';
  tableRef: RefObject<HTMLElement>;
  onSubmit?: (type: FileFormat) => void;
}

const DownloadTable = ({
  fileName,
  fullTable,
  headingTag = 'h3',
  headingSize = 's',
  tableRef,
  onSubmit,
}: Props) => {
  return (
    <Formik<FormValues>
      initialValues={{
        fileFormat: undefined,
      }}
      onSubmit={values => {
        if (onSubmit) {
          onSubmit(values.fileFormat);
        }
        return values.fileFormat === 'csv'
          ? handleCsvDownload(fileName, fullTable)
          : handleOdsDownload(fileName, fullTable.subjectMeta, tableRef);
      }}
      validationSchema={Yup.object<FormValues>({
        fileFormat: Yup.mixed().required('Choose a file format'),
      })}
    >
      {() => {
        return (
          <Form id="downloadTableForm">
            <>
              {createElement(
                headingTag,
                {
                  className: `govuk-heading-${headingSize}`,
                },
                'Download Table',
              )}
              <FormFieldRadioGroup<FormValues>
                legend="Select file format:"
                legendSize="s"
                legendWeight="regular"
                name="fileFormat"
                small
                order={[]}
                options={[
                  {
                    label:
                      'Table in ODS format (spreadsheet, with title and footnotes)',
                    value: 'ods',
                  },
                  {
                    label:
                      'Table in CSV format (flat file, with location codes)',
                    value: 'csv',
                  },
                ]}
              />
              <Button type="submit">Download table</Button>
            </>
          </Form>
        );
      }}
    </Formik>
  );
};

export default DownloadTable;
