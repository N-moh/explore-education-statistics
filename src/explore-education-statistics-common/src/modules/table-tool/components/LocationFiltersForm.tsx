import { Form, FormFieldset, Formik } from '@common/components/form';
import SummaryList from '@common/components/SummaryList';
import SummaryListItem from '@common/components/SummaryListItem';
import Yup from '@common/lib/validation/yup';
import { PublicationSubjectMeta } from '@common/modules/full-table/services/tableBuilderService';
import { Dictionary } from '@common/types/util';
import useResetFormOnPreviousStep from '@common/modules/table-tool/components/hooks/useResetFormOnPreviousStep';
import { FormikProps } from 'formik';
import sortBy from 'lodash/sortBy';
import React, { useRef } from 'react';
import { useImmer } from 'use-immer';
import FormFieldCheckboxMenu from './FormFieldCheckboxMenu';
import { InjectedWizardProps } from './Wizard';
import WizardStepFormActions from './WizardStepFormActions';
import WizardStepHeading from './WizardStepHeading';

export type LocationsFormValues = Dictionary<string[]>;

interface FormValues {
  locations: LocationsFormValues;
}

export type LocationFiltersFormSubmitHandler = (values: FormValues) => void;

interface Props {
  options: PublicationSubjectMeta['locations'];
  onSubmit: LocationFiltersFormSubmitHandler;
}

const LocationFiltersForm = (props: Props & InjectedWizardProps) => {
  const {
    options,
    onSubmit,
    isActive,
    goToNextStep,
    currentStep,
    stepNumber,
  } = props;

  const formikRef = useRef<Formik<FormValues>>(null);
  const formId = 'locationFiltersForm';

  const [locationLevels, updateLocationLevels] = useImmer<
    Dictionary<{ label: string; value: string }[]>
  >({});

  useResetFormOnPreviousStep(formikRef, currentStep, stepNumber, () => {
    updateLocationLevels(() => {
      return {};
    });
  });

  const stepHeading = (
    <WizardStepHeading {...props} fieldsetHeading>
      Choose locations
    </WizardStepHeading>
  );

  return (
    <Formik<FormValues>
      enableReinitialize
      ref={formikRef}
      onSubmit={async values => {
        const locations = Object.entries(values.locations).reduce(
          (acc, [level, levelOptions]) => {
            if (levelOptions.length === 0) {
              return acc;
            }

            return {
              ...acc,
              [level]: levelOptions,
            };
          },
          {},
        );

        await onSubmit({ locations });
        goToNextStep();
      }}
      initialValues={{
        locations: Object.keys(options).reduce((acc, level) => {
          return {
            ...acc,
            [level]: [],
          };
        }, {}),
      }}
      validationSchema={Yup.object<FormValues>({
        locations: Yup.mixed().test(
          'required',
          'Select at least one location',
          (value: Dictionary<string[]>) =>
            Object.values(value).some(groupOptions => groupOptions.length > 0),
        ),
      })}
      render={(form: FormikProps<FormValues>) => {
        return isActive ? (
          <Form {...form} id={formId}>
            <FormFieldset
              id={`${formId}-levels`}
              legend={stepHeading}
              hint="Select at least one"
              error={
                typeof form.errors.locations === 'string'
                  ? form.errors.locations
                  : ''
              }
            >
              <div className="govuk-grid-row">
                <div className="govuk-grid-column-one-half-from-desktop">
                  {Object.entries(options).map(([levelKey, level]) => {
                    return (
                      <FormFieldCheckboxMenu
                        name={`locations.${levelKey}`}
                        key={levelKey}
                        options={level.options}
                        id={`${formId}-levels-${levelKey}`}
                        legend={level.legend}
                        legendHidden
                        selectAll={false}
                        onAllChange={() => {
                          updateLocationLevels(draft => {
                            draft[levelKey] =
                              draft[levelKey].length < level.options.length
                                ? level.options
                                : [];
                          });
                        }}
                        onChange={(event, option) => {
                          updateLocationLevels(draft => {
                            if (!draft[levelKey]) {
                              draft[levelKey] = [];
                            }

                            const { value } = event.target;

                            const matchingOption = draft[levelKey].find(
                              levelOption => levelOption.value === value,
                            );

                            if (matchingOption) {
                              draft[levelKey] = draft[levelKey].filter(
                                levelOption => levelOption.value !== value,
                              );
                            } else {
                              draft[levelKey].push({
                                value,
                                label: option.label,
                              });
                            }
                          });
                        }}
                      />
                    );
                  })}
                </div>
              </div>
            </FormFieldset>

            <WizardStepFormActions {...props} form={form} formId={formId} />
          </Form>
        ) : (
          <>
            {stepHeading}
            <SummaryList noBorder>
              {Object.entries(locationLevels)
                .filter(([_, levelOptions]) => levelOptions.length > 0)
                .map(([levelKey, levelOptions]) => (
                  <SummaryListItem
                    term={options[levelKey].legend}
                    key={levelKey}
                  >
                    {sortBy(levelOptions, ['label']).map(level => (
                      <React.Fragment key={level.value}>
                        {level.label}
                        <br />
                      </React.Fragment>
                    ))}
                  </SummaryListItem>
                ))}
            </SummaryList>
          </>
        );
      }}
    />
  );
};

export default LocationFiltersForm;
