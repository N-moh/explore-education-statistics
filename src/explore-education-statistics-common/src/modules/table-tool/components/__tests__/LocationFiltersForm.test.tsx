import LocationFiltersForm, {
  LocationFormValues,
} from '@common/modules/table-tool/components/LocationFiltersForm';
import { InjectedWizardProps } from '@common/modules/table-tool/components/Wizard';
import { SubjectMeta } from '@common/services/tableBuilderService';
import { waitFor } from '@testing-library/dom';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import noop from 'lodash/noop';
import React from 'react';

describe('LocationFiltersForm', () => {
  const testWizardStepProps: InjectedWizardProps = {
    currentStep: 1,
    isActive: true,
    isLoading: false,
    stepNumber: 1,
    setCurrentStep: noop,
    goToNextStep: noop,
    goToPreviousStep: noop,
    shouldScroll: false,
  };

  const testLocations: SubjectMeta['locations'] = {
    Country: {
      hint: '',
      legend: 'Country',
      options: [
        {
          label: 'Country 1',
          value: 'country-1',
        },
      ],
    },
    LocalAuthority: {
      hint: '',
      legend: 'Local authority',
      options: [
        {
          label: 'Local authority 1',
          value: 'local-authority-1',
        },
        {
          label: 'Local authority 2',
          value: 'local-authority-2',
        },
        {
          label: 'Local authority 3',
          value: 'local-authority-3',
        },
      ],
    },
    Region: {
      hint: '',
      legend: 'Region',
      options: [
        {
          label: 'Region 1',
          value: 'region-1',
        },
        {
          label: 'Region 2',
          value: 'region-2',
        },
      ],
    },
  };

  const testSingleLocation: SubjectMeta['locations'] = {
    Country: {
      hint: '',
      legend: 'Country',
      options: [
        {
          label: 'Country 1',
          value: 'country-1',
        },
      ],
    },
  };

  test('renders location group options correctly', () => {
    render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testLocations}
        onSubmit={noop}
      />,
    );

    expect(
      within(
        screen.getByRole('group', {
          name: 'Country',
          hidden: true,
        }),
      ).getAllByRole('checkbox', {
        hidden: true,
      }),
    ).toHaveLength(1);

    expect(
      within(
        screen.getByRole('group', {
          name: 'Local authority',
          hidden: true,
        }),
      ).getAllByRole('checkbox', {
        hidden: true,
      }),
    ).toHaveLength(3);

    expect(
      within(
        screen.getByRole('group', {
          name: 'Region',
          hidden: true,
        }),
      ).getAllByRole('checkbox', {
        hidden: true,
      }),
    ).toHaveLength(2);
  });

  test('selecting options shows the number of selected options for each location group', () => {
    render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testLocations}
        onSubmit={noop}
      />,
    );

    userEvent.click(screen.getByLabelText('Local authority 2'));

    expect(
      screen.getByRole('button', {
        name: 'Local authority - 1 selected',
      }),
    ).toBeInTheDocument();

    userEvent.click(screen.getByLabelText('Local authority 3'));
    userEvent.click(screen.getByLabelText('Country 1'));
    userEvent.click(screen.getByLabelText('Region 2'));

    expect(
      screen.getByRole('button', {
        name: 'Local authority - 2 selected',
      }),
    ).toBeInTheDocument();

    expect(
      screen.getByRole('button', {
        name: 'Country - 1 selected',
      }),
    ).toBeInTheDocument();

    expect(
      screen.getByRole('button', {
        name: 'Region - 1 selected',
      }),
    ).toBeInTheDocument();
  });

  test('shows validation errors if no location is selected', async () => {
    render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testLocations}
        onSubmit={noop}
      />,
    );

    userEvent.click(
      screen.getByRole('button', {
        name: 'Next step',
      }),
    );

    await waitFor(() => {
      expect(
        screen.getByRole('link', {
          name: 'Select at least one location',
        }),
      ).toBeInTheDocument();
    });
  });

  test('only the correct checkboxes are selected from initial values', () => {
    render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testLocations}
        onSubmit={noop}
        initialValues={{
          LocalAuthority: ['local-authority-1', 'local-authority-3'],
          Region: ['region-2'],
        }}
      />,
    );

    expect(screen.getByLabelText('Local authority 1')).toHaveAttribute(
      'checked',
    );
    expect(screen.getByLabelText('Local authority 3')).toHaveAttribute(
      'checked',
    );
    expect(screen.getByLabelText('Region 2')).toHaveAttribute('checked');

    expect(screen.getByLabelText('Country 1')).not.toHaveAttribute('checked');
    expect(screen.getByLabelText('Local authority 2')).not.toHaveAttribute(
      'checked',
    );
    expect(screen.getByLabelText('Region 1')).not.toHaveAttribute('checked');
  });

  test('automatically selects the option and expands the group if only one location available', () => {
    render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testSingleLocation}
        onSubmit={noop}
      />,
    );

    expect(
      screen.getByRole('button', {
        name: 'Country - 1 selected',
      }),
    ).toBeInTheDocument();

    expect(
      within(
        screen.getByRole('group', {
          name: 'Country',
        }),
      ).getAllByRole('checkbox'),
    ).toHaveLength(1);

    expect(screen.getByLabelText('Country 1')).toBeVisible();

    expect(screen.getByLabelText('Country 1')).toHaveAttribute('checked');
  });

  test('renders a read-only view of selected options when no longer the current step', async () => {
    const { container, rerender } = render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testLocations}
        onSubmit={noop}
      />,
    );

    userEvent.click(screen.getByLabelText('Region 1'));
    userEvent.click(screen.getByLabelText('Region 2'));
    userEvent.click(screen.getByLabelText('Country 1'));

    await rerender(
      <LocationFiltersForm
        {...testWizardStepProps}
        isActive={false}
        options={testLocations}
        onSubmit={noop}
      />,
    );

    expect(screen.queryAllByRole('checkbox')).toHaveLength(0);
    expect(container.querySelector('dl')).toMatchSnapshot();
  });

  test('clicking `Next step` calls `onSubmit` with correct values', async () => {
    const handleSubmit = jest.fn();
    render(
      <LocationFiltersForm
        {...testWizardStepProps}
        options={testLocations}
        onSubmit={handleSubmit}
      />,
    );

    userEvent.click(screen.getByLabelText('Local authority 3'));
    userEvent.click(screen.getByLabelText('Region 1'));

    userEvent.click(screen.getByRole('button', { name: 'Next step' }));

    const expected: LocationFormValues = {
      locations: {
        LocalAuthority: ['local-authority-3'],
        Region: ['region-1'],
      },
    };

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(1);
      expect(handleSubmit).toHaveBeenCalledWith(expected);
    });
  });
});
