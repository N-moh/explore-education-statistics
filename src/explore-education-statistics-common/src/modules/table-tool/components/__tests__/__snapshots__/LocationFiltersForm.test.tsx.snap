// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationFiltersForm renders a read-only view of selected options when no longer the current step 1`] = `
<dl
  class="govuk-summary-list govuk-summary-list--no-border"
>
  <div
    class="govuk-summary-list__row"
    data-testid="Country"
  >
    <dt
      class="govuk-summary-list__key"
      data-testid="Country-key"
    >
      Country
    </dt>
    <dd
      class="govuk-summary-list__value"
      data-testid="Country-value"
    >
      <ul
        class="govuk-list listContainer"
      >
        <li>
          Country 1
        </li>
      </ul>
    </dd>
  </div>
  <div
    class="govuk-summary-list__row"
    data-testid="Region"
  >
    <dt
      class="govuk-summary-list__key"
      data-testid="Region-key"
    >
      Region
    </dt>
    <dd
      class="govuk-summary-list__value"
      data-testid="Region-value"
    >
      <ul
        class="govuk-list listContainer"
      >
        <li>
          Region 1
        </li>
        <li>
          Region 2
        </li>
      </ul>
    </dd>
  </div>
</dl>
`;
